alibi_detect/od/pytorch/base.py

```diff
@@ -0,0 +1,239 @@
+from __future__ import annotations
+from typing import List, Union, Optional, Dict
+from dataclasses import dataclass, asdict
+from abc import ABC, abstractmethod
+from functools import singledispatch
+
+
+import numpy as np
+import torch
+
+from alibi_detect.od.pytorch.ensemble import FitMixinTorch
+from alibi_detect.utils.pytorch.misc import get_device
+from alibi_detect.od.base import ThresholdNotInferredException
+
+
+@dataclass
+class TorchOutlierDetectorOutput:
+    """Output of the outlier detector."""
+    threshold_inferred: bool
+    instance_score: torch.Tensor
+    threshold: Optional[torch.Tensor]
+    is_outlier: Optional[torch.Tensor]
+    p_value: Optional[torch.Tensor]
+
+
+@singledispatch
+def to_numpy(arg):
+    """Converts any `torch` tensors found in input to `numpy` arrays.
+
+    Takes a `torch` tensor or `TorchOutlierDetectorOutput` and converts any `torch` tensors found to `numpy` arrays
+
+    Parameters
+    ----------
+    x
+        Data to convert.
+
+    Returns
+    -------
+    `np.ndarray` or dictionary of containing `numpy` arrays
+    """
+    raise NotImplementedError(f"Cannot transform type {type(arg)} to numpy array.")
+
+
+@to_numpy.register
+def _(x: torch.Tensor) -> np.ndarray:
+    return x.cpu().detach().numpy()
```

Comment:
I generally like this `singledispatch` pattern, it seems quite elegant. 

As an aside related to the `.detach()`. We could look at wrapping pretty much the entire torch backend in `with torch.no_grad()` for a (probably minor) performance gain?  As you said offline though @mauicv, we'd still need the `.detach()` here since some detectors do use gradients.

Reply:
I've added `@torch.no_grad` to the relevant method calls!

Reply:
Ok so just to check my understanding. `KNNTorch.score` is now wrapped in `torch.no_grad`, but we are keeping the `detach()` here because other detectors `score` methods will use gradients, hence won't be wrapped in `torch.no_grad`?